x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/action/filter/ShardBulkInferenceActionFilter.java

```diff
@@ -0,0 +1,534 @@
+                for (var entry : inferenceRequests.entrySet()) {
+                    executeShardBulkInferenceAsync(entry.getKey(), null, entry.getValue(), releaseOnFinish.acquire());
+                }
+            }
+        }
+
+        private void executeShardBulkInferenceAsync(
+            final String inferenceId,
+            @Nullable InferenceProvider inferenceProvider,
+            final List<FieldInferenceRequest> requests,
+            final Releasable onFinish
+        ) {
+            if (inferenceProvider == null) {
+                ActionListener<ModelRegistry.UnparsedModel> modelLoadingListener = new ActionListener<>() {
+                    @Override
+                    public void onResponse(ModelRegistry.UnparsedModel unparsedModel) {
+                        var service = inferenceServiceRegistry.getService(unparsedModel.service());
+                        if (service.isEmpty() == false) {
+                            var provider = new InferenceProvider(
+                                service.get(),
+                                service.get()
+                                    .parsePersistedConfigWithSecrets(
+                                        inferenceId,
+                                        unparsedModel.taskType(),
+                                        unparsedModel.settings(),
+                                        unparsedModel.secrets()
+                                    )
+                            );
+                            executeShardBulkInferenceAsync(inferenceId, provider, requests, onFinish);
+                        } else {
+                            try (onFinish) {
+                                for (int i = 0; i < requests.size(); i++) {
+                                    var request = requests.get(i);
+                                    inferenceResults.get(request.index).failures.add(
+                                        new ResourceNotFoundException(
+                                            "Inference service [{}] not found for field [{}]",
+                                            unparsedModel.service(),
+                                            request.field
+                                        )
+                                    );
+                                }
+                            }
+                        }
+                    }
+
+                    @Override
+                    public void onFailure(Exception exc) {
+                        try (onFinish) {
+                            for (int i = 0; i < requests.size(); i++) {
+                                var request = requests.get(i);
+                                inferenceResults.get(request.index).failures.add(
+                                    new ResourceNotFoundException("Inference id [{}] not found for field [{}]", inferenceId, request.field)
+                                );
+                            }
+                        }
```

Comment:
It seems to me that ResourceNotFoundException isn't the only thing that could cause a failure here. 

Could you unwrap and check the exception to ensure we should return this error? Otherwise we should bubble up the exception so that folks know whats going on.

Reply:
I see your point. I've added that in 5dc093d07e68f1488c457fb62b531207ef7d8bce

 Maybe we should just wrap the original exception into an `ElasticsearchException`, saying that we had a problem in field x? I'm not sure about adding a `ResourceNotFoundException` as a result of the inference not being present - it would be more a 400 IMO, as the user is providing an inference that can't be used for whatever reason but the action resource (the doc) exists...

cc @jimczi for thoughts here.